core/primitives/src/bandwidth_scheduler.rs

```diff
@@ -79,6 +79,54 @@ pub struct BandwidthRequest {
     pub requested_values_bitmap: BandwidthRequestBitmap,
 }
 
+impl BandwidthRequest {
+    /// Creates a bandwidth request based on the sizes of receipts in the outgoing buffer.
+    /// Returns None when a request is not needed (receipt size below base bandwidth).
+    pub fn make_from_receipt_sizes(
+        to_shard: u8,
+        receipt_sizes: impl Iterator<Item = u64>,
+        params: &BandwidthSchedulerParams,
+    ) -> Option<BandwidthRequest> {
+        let values = BandwidthRequestValues::new(params).values;
+        let mut bitmap = BandwidthRequestBitmap::new();
+
+        // For every receipt find out how much bandwidth would be needed to send out
+        // all the receipts up to this one. Then find the value that is at least as
+        // large as the required bandwidth and request it in the request bitmap.
+        let mut total_size: u64 = 0;
+        let mut cur_value_idx: usize = 0;
```

Comment:
hey, the two pointers algorithm from the interviews is useful for something after all :)